src/framework/pybind_basics.hpp

```diff
@@ -0,0 +1,172 @@
+// Move a matrix to Python via conversion to Numpy array
+template <typename T> py::object to_python(matrix<T> &&obj);
+
+// Move a Vector to Python via conversion to Numpy array
+template <typename T> py::object to_python(AER::Vector<T> &&obj);
+
+// Move a Vector to Python via recusivly calling to_python on elements
+template <typename T> py::object to_python(std::vector<T> &&obj);
+
+// Template specialization for moving numeric std::vectors to Numpy arrays
+template <> py::object to_python(std::vector<AER::int_t> &&obj);
+template <> py::object to_python(std::vector<AER::uint_t> &&obj);
+template <> py::object to_python(std::vector<float> &&obj);
+template <> py::object to_python(std::vector<double> &&obj);
+template <> py::object to_python(std::vector<std::complex<double>> &&obj);
+template <> py::object to_python(std::vector<std::complex<float>> &&obj);
+
+//------------------------------------------------------------------------------
+// Convert To Numpy Arrays
+//------------------------------------------------------------------------------
+
+// Convert a matrix to a 2D Numpy array in Fortan order
+template <typename T>
+py::array_t<T, py::array::f_style> to_numpy(matrix<T> &&obj);
+
+// Convert a Vector to a 1D Numpy array
+template <typename T>
+py::array_t<T> to_numpy(AER::Vector<T> &&obj);
+
+// Convert a vector to a 1D Numpy array
+template <typename T>
+py::array_t<T> to_numpy(std::vector<T> &&obj);
+
+//============================================================================
+// Implementation
+//============================================================================
+
+//------------------------------------------------------------------------------
+// Basic Types
+//------------------------------------------------------------------------------
+
+template <typename T>
+py::object to_python(T &&obj) {
+  return py::cast(obj, py::return_value_policy::move);
+}
+
+template <typename T>
+py::object to_python(std::vector<T> &&obj) {
+  py::list pydata;
+  for(auto& elt : obj) {
+    pydata.append(to_python(std::move(elt)));
+  }
+  return std::move(pydata);
+}
+
+template <typename T>
+py::object to_python(matrix<T> &&obj) {
+  return to_numpy(std::move(obj));
+}
+
+template <typename T>
+py::object to_python(AER::Vector<T> &&obj) {
+  return to_numpy(std::move(obj));
+}
+
+template <>
+py::object to_python(std::vector<AER::int_t> &&obj) {
+  return to_numpy(std::move(obj));
+}
+
+template <>
+py::object to_python(std::vector<AER::uint_t> &&obj) {
+  return to_numpy(std::move(obj));
+}
+
+template <>
+py::object to_python(std::vector<double> &&obj) {
+  return to_numpy(std::move(obj));
+}
+
+template <>
+py::object to_python(std::vector<float> &&obj) {
+  return to_numpy(std::move(obj));
+}
+
+template <>
+py::object to_python(std::vector<std::complex<double>> &&obj) {
+  return to_numpy(std::move(obj));
+}
+
+template <>
+py::object to_python(std::vector<std::complex<float>> &&obj) {
+  return to_numpy(std::move(obj));
+}
```

Comment:
I don't want to overload all std::vectors to go to a numpy array. Using a specialization means a std::vector<T> that doesn't match one of these templates will be converted to a python list instead of a numpy array

Reply:
Oh, right! My bad, I confused `AER::Vector<T>` with `std::vector<T>` in a function above. I guess I need more coffee on mornings... ;)